src/agora/dodemul.cpp

```diff
@@ -161,9 +161,13 @@ Event_data DoDemul::launch(size_t tag)
             mkl_jit_cgemm(jitter, (MKL_Complex8*)ul_zf_ptr,
                 (MKL_Complex8*)data_ptr, (MKL_Complex8*)equal_ptr);
 #else
-            const cx_fmat mat_data(data_ptr, cfg->BS_ANT_NUM, 1, false);
-            const cx_fmat mat_ul_zf(
-                ul_zf_ptr, cfg->UE_NUM, cfg->BS_ANT_NUM, false);
+            cx_fmat mat_data(data_ptr, cfg->BS_ANT_NUM, 1, false);
+
+            cx_fmat mat_ul_zf(ul_zf_ptr, cfg->UE_NUM, cfg->BF_ANT_NUM, false);
+            if (cfg->external_ref_node) {
+                mat_ul_zf.insert_cols(cfg->ref_ant,
+                    cx_fmat(cfg->UE_NUM, cfg->nChannels, fill::zeros));
+            }
```

Comment:
I am a little confused. When using an external node, should we collect data symbols on that node (should` mat_data` have size `BF_ANT_NUMx1` or `BS_ANT_NUMx1`)?

Should we move line 168-169 here and line 170-171 in doprecode.cpp to dozf.cpp? By moving them to dozf.cpp, I think we can just use BS_ANT_NUM in agora.cpp to make it less confusing.

Reply:
I think the reason I did not did everything in `dozf.cpp` was to spread the computation across threads. But I have not benchmarked `shed_row` and `insert_cols` functions, so it might not matter.

Reply:
Also right now we treat external ref antenna the same as other in terms of receiving uplink and transmitting downlink, in order to keep the part of the code from `dofft` and below the same for both external and internal ref nodes. In this case `mat_data` above has the size `BS_ANT_NUM`. So before we do demultiplexing we add a zero column to `mat_dl_zf` to make it the same size in case of external ref node. Similarly for downlink in `doprecode`.

If we were to change this perhaps we should do similar things in `dofft`. So it is choice we need to make.

Reply:
I moved all the logic related to `external_ref_node` to dozf.